README.md
# flutter_avatar

Various ready to use Avatars in your Flutter apps

<p align="center">
<img src="https://raw.githubusercontent.com/timilehinjegede/flutter-avatar/master/screenshots/fa_1.png" alt="Showcase 1" width="230px" hspace="30"/>  <img src="https://raw.githubusercontent.com/timilehinjegede/flutter-avatar/master/screenshots/fa_2.png" alt="Showcase 2" width="230px"/> 
</p>

## Installing

In the `pubspec.yaml` of your flutter project, add the following dependency:

```yaml
dependencies:
  flutter_avatar: ^0.1.0
```

## How To Use

In your file add the following import:

```dart
import 'package:flutter_avatar/flutter_avatar.dart';
```

Create a `FlutterAvatar` widget:

```dart
  FlutterAvatar(
      backgroundColor: Colors.blue,
      radius: 50,
      shape: BoxShape.circle,
      avatarShadow: BoxShadow(
        color: Colors.blue.withOpacity(0.6),
        blurRadius: 15,
        spreadRadius: 5,
        offset: Offset(0, -1),
      ),
      child: Text(
        'TJ',
        style: TextStyle(
          fontSize: 35,
          color: Colors.white,
          fontWeight: FontWeight.bold,
        ),
      ),
    );
```

## Customizations 

Property | Description
-------- | ---------------
radius    | The size of the [FlutterAvatar], which is half of the Diameter. Defaults to 20.
child    | The widget below this widget [FlutterAvatar] in the widget tree
borderRadius    | The corners of the [FlutterAvatar] are rounded by the value of the [borderRadius].
shape    | The shape to be used for the [FlutterAvatar], can be [BoxShape.circle] or [BoxShape.rectangle].
badge    | The title displayed to the user
badgePosition    | The title displayed to the user
avatarShadow    | The [avatarShadow] is cast behind the [FlutterAvatar] widget
backgroundColor    | The color used to fill the [FlutterAvatar]
elevation    | The elevation of the [FlutterAvatar] widget. Defaults to 5.0 if no value is given.
elevationColor    | The color of the [elevation]


## Contributions 

Feel free to contribute to this project.

If you find a bug or want a feature, but don't know how to fix/implement it, please fill an [issue](https://github.com/timilehinjegede/flutter-avatar/issues).  
If you fixed a bug or implemented a feature, please send a [pull request](https://github.com/timilehinjegede/flutter-avatar/pulls).

CHANGELOG.md
## [0.1.0] - 2020-11-01

* Initial beta release.


test/flutter_avatar_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_avatar/flutter_avatar.dart';

void main() {
  test('adds one to input values', () {});
}


lib/flutter_avatar.dart
library flutter_avatar;

import 'package:flutter/material.dart';

enum BadgePosition { bottomRight, bottomLeft, topRight, topLeft }

class FlutterAvatar extends StatelessWidget {
  /// Creates a avatar widget which is highly customizable
  const FlutterAvatar({
    this.radius = 20.0,
    this.borderRadius,
    this.shape = BoxShape.rectangle,
    this.badgePosition,
    this.badge,
    this.avatarShadow,
    this.backgroundColor,
    this.child,
    this.elevation = 5.0,
    this.elevationColor = Colors.black,
  }) : assert(shape != BoxShape.circle || borderRadius == null);

  /// The size of the [FlutterAvatar], which is half of the Diameter
  ///
  /// defaults to 20
  final double radius;

  /// if non-null, the corners of the [FlutterAvatar] are rounded by the value of the [borderRadius]
  ///
  /// Can only be used when [shape] is [BosShape.rectangle], cannot be used if [shape] is [BoxShape.circle].
  final BorderRadius borderRadius;

  /// The shape to be used for the [FlutterAvatar].
  ///
  /// If [shape] is [BoxShape.circle], the [borderRadius] cannot be used.
  /// If [shape] is [BoxShape.rectangle], specify a [borderRadius].
  final BoxShape shape;

  /// Used to change the position of the [badge] widget that is stacked on top of [FlutterAvatar]
  ///
  /// defaults to [BadgePosition.topLeft]
  final BadgePosition badgePosition;

  /// The widget that is stacked on top of the [FlutterAvatar]
  ///
  /// Use [badgePosition] to change the position of the [badge] widget.
  final Widget badge;

  /// The color used to fill the [FlutterAvatar].
  final Color backgroundColor;

  /// The widget below this widget [FlutterAvatar] in the widget tree
  ///
  /// Can be a [Text] widget. If the [FlutterAvatar] requires an image, use [backgroundImage].
  final Widget child;

  /// The [avatarShadow] is cast behind the [FlutterAvatar] widget
  final BoxShadow avatarShadow;

  /// The elevation of the [FlutterAvatar] widget.
  ///
  /// The [elevation] defaults to 5.0 if a value isn't specified.
  final double elevation;

  /// The color of the [elevation]
  final Color elevationColor;

  @override
  Widget build(BuildContext context) {
    /// returns the [_ItemWidget] if the [badge] is null
    return badge == null
        ? _ItemWidget(
            radius: radius,
            borderRadius: borderRadius,
            shape: shape,
            color: backgroundColor,
            boxShadow: avatarShadow,
            elevation: elevation,
            elevationColor: elevationColor,
            child: child,
            // image: widget.backgroundImage,
          )

        /// returns a stack of widgets containing the [_ItemWidget] andd the [badge] if the [badge] is specified
        : SizedBox(
            height: radius * 2,
            width: radius * 2,
            child: Stack(
              children: [
                Center(
                  child: _ItemWidget(
                    radius: radius,
                    borderRadius: borderRadius,
                    shape: shape,
                    color: backgroundColor,
                    boxShadow: avatarShadow,
                    elevation: elevation,
                    elevationColor: elevationColor,
                    child: child,
                    // image: widget.backgroundImage,
                  ),
                ),
                Positioned(
                  left: _mapPositionToValues(badgePosition)[0],
                  top: _mapPositionToValues(badgePosition)[1],
                  right: _mapPositionToValues(badgePosition)[2],
                  bottom: _mapPositionToValues(badgePosition)[3],
                  child: badge,
                  // alignment: badgePosition,
                ),
              ],
            ),
          );
  }

  /// used to determine the [badgePosition] to be used for the [badge]
  List<double> _mapPositionToValues(BadgePosition badgePosition) {
    // [left value, top value, right value, bottom value]
    switch (badgePosition) {
      case BadgePosition.bottomRight:
        // return array containing bottom and right values
        return [null, null, 0.0, 0.0];
      case BadgePosition.bottomLeft:
        return [0.0, null, null, 0.0];
      case BadgePosition.topRight:
        return [null, 0.0, 0.0, null];
      case BadgePosition.topLeft:
        return [0.0, 0.0, null, null];
      default:
        // topLeft for default
        return [null, 0.0, 0.0, null];
    }
  }
}

/// creates an [_ItemWidget]
class _ItemWidget extends StatelessWidget {
  const _ItemWidget({
    Key key,
    this.radius,
    this.borderRadius,
    this.shape,
    this.color,
    this.enableBlur,
    this.blurColor,
    this.child,
    this.boxShadow,
    this.elevation,
    this.elevationColor,
  }) : super(key: key);

  final double radius;
  final BorderRadius borderRadius;
  final BoxShape shape;
  final Color color;
  final Widget child;
  final bool enableBlur;
  final Color blurColor;
  final BoxShadow boxShadow;
  final double elevation;
  final Color elevationColor;

  @override
  Widget build(BuildContext context) {
    return Material(
      elevation: elevation,
      color: Colors.transparent,
      type: borderRadius == null ? MaterialType.circle : MaterialType.canvas,
      borderRadius: borderRadius,
      shadowColor: elevationColor,
      child: SizedBox(
        height: radius * 2,
        width: radius * 2,
        child: DecoratedBox(
          decoration: BoxDecoration(
            borderRadius: borderRadius,
            shape: shape,
            color: color,
            boxShadow: boxShadow != null ? [boxShadow] : null,
          ),
          child: Center(
            child: child,
          ),
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_avatar/flutter_avatar.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Flutter Avatar'),
    );
  }
}

class MyHomePage extends StatelessWidget {
  final String title;

  const MyHomePage({Key key, this.title}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(title),
      ),
      body: SingleChildScrollView(
        child: Padding(
          padding: const EdgeInsets.only(top: 25.0),
          child: SizedBox(
            width: double.infinity,
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.center,
              children: [
                // Example 1
                FAExampleOne(),

                const SizedBox(
                  height: 35,
                ),

                // Example 2
                FAExampleTwo(),

                const SizedBox(
                  height: 35,
                ),

                // Example 3
                FAExampleThree(),

                const SizedBox(
                  height: 35,
                ),

                // Example 4
                FAExampleFour(),
              ],
            ),
          ),
        ),
      ),
    );
  }
}

class FAExampleOne extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return FlutterAvatar(
      backgroundColor: Colors.red,
      radius: 50,
      borderRadius: BorderRadius.all(Radius.circular(30)),
      badge: Container(
        height: 25,
        width: 25,
        decoration: BoxDecoration(
          shape: BoxShape.circle,
          color: Colors.green,
        ),
        child: Icon(
          Icons.add,
          size: 20,
          color: Colors.white,
        ),
      ),
      badgePosition: BadgePosition.topRight,
      avatarShadow: BoxShadow(
        color: Colors.purple.withOpacity(0.6),
        blurRadius: 20,
        spreadRadius: 5,
        offset: Offset(0, 2),
      ),
      child: Text(
        'TJ',
        style: TextStyle(
          fontSize: 35,
          color: Colors.white,
          fontWeight: FontWeight.bold,
        ),
      ),
    );
  }
}

class FAExampleTwo extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return FlutterAvatar(
      backgroundColor: Colors.blue,
      radius: 50,
      shape: BoxShape.circle,
      avatarShadow: BoxShadow(
        color: Colors.blue.withOpacity(0.6),
        blurRadius: 15,
        spreadRadius: 5,
        offset: Offset(0, -1),
      ),
      child: Text(
        'TJ',
        style: TextStyle(
          fontSize: 35,
          color: Colors.white,
          fontWeight: FontWeight.bold,
        ),
      ),
    );
  }
}

class FAExampleThree extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return FlutterAvatar(
      backgroundColor: Colors.purple,
      radius: 50,
      shape: BoxShape.circle,
      badge: Container(
        height: 25,
        width: 25,
        decoration: BoxDecoration(
          shape: BoxShape.circle,
          color: Colors.grey,
        ),
        child: Icon(
          Icons.photo_camera,
          size: 15,
          color: Colors.white,
        ),
      ),
      badgePosition: BadgePosition.bottomRight,
      avatarShadow: BoxShadow(
        color: Colors.purple.withOpacity(0.6),
        blurRadius: 20,
        spreadRadius: 5,
        offset: Offset(0, 2),
      ),
      child: Text(
        'TJ',
        style: TextStyle(
          fontSize: 35,
          color: Colors.white,
          fontWeight: FontWeight.bold,
        ),
      ),
    );
  }
}

class FAExampleFour extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return FlutterAvatar(
      backgroundColor: Colors.green,
      radius: 50,
      shape: BoxShape.circle,
      elevation: 10,
      child: Text(
        'TJ',
        style: TextStyle(
          fontSize: 35,
          color: Colors.white,
          fontWeight: FontWeight.bold,
        ),
      ),
    );
  }
}
